src/transports/eip_1193.rs

```diff
@@ -0,0 +1,189 @@
+                    let copy = self.provider.clone();
+                    Box::pin(async move {
+                        copy.request_wrapped(RequestArguments {
+                            method: method_call.method,
+                            params: js_params,
+                        })
+                        .await
+                    })
+                }
+                _other => Box::pin(future::err(Error::Internal)),
+            },
+            _other => Box::pin(future::err(Error::Internal)),
+        }
+    }
+}
+
+impl DuplexTransport for Eip1193 {
+    type NotificationStream = mpsc::UnboundedReceiver<serde_json::Value>;
+
+    fn subscribe(&self, id: SubscriptionId) -> error::Result<Self::NotificationStream> {
+        let (sender, receiver) = mpsc::unbounded();
+        let mut subscriptions_ref = self.subscriptions.borrow_mut();
+        subscriptions_ref.insert(id, sender);
+        Ok(receiver)
+    }
+
+    fn unsubscribe(&self, id: SubscriptionId) -> error::Result<()> {
+        match (*self.subscriptions.borrow_mut()).remove(&id) {
+            Some(_sender) => Ok(()),
+            None => Err(Error::Internal),
+        }
+    }
+}
+
+#[wasm_bindgen]
+// Rustfmt removes the 'async' keyword from async functions in extern blocks. It's fixed
+// in rustfmt 2.
+#[rustfmt::skip]
+extern "C" {
+    #[wasm_bindgen]
+    #[derive(Clone, Debug)]
+    /// An EIP-1193 provider object. Available by convention at `window.ethereum`
+    pub type Provider;
+
+    #[wasm_bindgen(catch, method)]
+    async fn request(_: &Provider, args: RequestArguments) -> Result<JsValue, JsValue>;
+
+    #[wasm_bindgen(method)]
+    fn on(_: &Provider, eventName: &str, listener: &Closure<dyn FnMut(JsValue)>);
+}
+
+impl Provider {
+    /// Get the provider at `window.ethereum`.
+    pub fn get_default() -> Result<Self, JsValue> {
+        get_provider_js()
+    }
+
+    async fn request_wrapped(&self, args: RequestArguments) -> error::Result<serde_json::value::Value> {
+        let js_result = self.request(args).await;
+        match js_result {
+            Ok(res) => Ok(res.into_serde().expect("couldn't translate request via JSON")),
+            Err(err) => {
+                match err.into_serde() {
+                    Ok(json_rpc_err) => Err(Error::Rpc(json_rpc_err)),
+                    // THE EIP says an error response MUST match the JSON RPC error structure.
+                    Err(_) => Err(Error::InvalidResponse(format!("{:?}", err))),
+                }
+            }
+        }
+    }
+}
+
+#[wasm_bindgen(inline_js = "export function get_provider_js() {return window.ethereum}")]
+extern "C" {
+    #[wasm_bindgen(catch)]
+    fn get_provider_js() -> Result<Provider, JsValue>;
+}
+
+#[wasm_bindgen]
+struct RequestArguments {
+    method: String,
+    params: js_sys::Array,
+}
+
+#[wasm_bindgen]
+impl RequestArguments {
+    #[wasm_bindgen(getter)]
+    pub fn method(&self) -> String {
+        self.method.clone()
```

Comment:
Do we need to clone? Why can't we return just references?

Reply:
You can't return references to JS-land unfortunately. There's no way for the borrow checker to see into JavaScript.